src/java_bytecode/generate_java_generic_type.cpp

```diff
@@ -115,8 +116,24 @@ typet generate_java_generic_typet::substitute_type(
                                   .type_variable()
                                   .get_identifier();
 
-    optionalt<size_t> results =
-      java_generics_get_index_for_subtype(generic_class, component_identifier);
+    // see if it is a generic parameter introduced by this class
+    optionalt<size_t> results;
+    if(is_java_generic_class_type(class_definition))
+    {
+      const java_generic_class_typet &generic_class =
+        to_java_generic_class_type(class_definition);
+
+      results = java_generics_get_index_for_subtype(
+        generic_class, component_identifier);
+    }
+    // see if it is an implicit generic parameter introduced by an outer class
+    else if(is_java_implicitly_generic_class_type(class_definition))
+    {
```

Comment:
Since it is either a `java_generic_class_type` or `is_java_implicitly_generic_class_type` (by above invariant) I think for clarity this should become an else with an invariant on the `is_java_implicitly_generic_class_type) as otherwise the below invariant (that `result.has_value()`) will confusingly report not finding the generic component.

Reply:
I'm also a bit concerned that these conditions appear to be orthogonal but in the code seems like they are mutually exclusive. I think this is a problem of the choice of using inheritance for all these properties instead of just properties of a `java_class_type`  and might come to bite us when we have a generic outer and generic inner class.

Reply:
They are not mutually exclusive. However, the generic parameter at hand must either be a generic type of the class or an implicit generic type coming from an outer class. I restructured the conditions here to reflect that correctly.